stgit/commands/rebase.py

```diff
@@ -61,11 +76,156 @@
 
 directory = DirectoryGotoTopLevel()
 
+INTERACTIVE_APPLY_LINE = '# --- APPLY_LINE ---'
+
+INTERACTIVE_HELP_INSTRUCTIONS = """
+# Commands:
+# k, keep <patch> = do not modify this patch
+# s, squash <patch> = squash patch into the previous patch
+# d, delete <patch> = delete patch
+#
+# These lines can be re-ordered; they are executed from top to bottom.
+#
+# Patches above the APPLY_LINE are applied; other patches are kept unapplied.
+#
+# If you remove everything the rebase will be aborted.
+"""
+
+
+def __get_description(stack, patch):
+    """Extract and return a patch's short description"""
+    cd = stack.patches[patch].data
+    return cd.message_str.strip().split('\n', 1)[0].rstrip()
+
+
+class Action(Enum):
+    # do nothing; a 'no-op'
+    KEEP = 0
+    # delete the patch
+    DELETE = 1
+    # squash the patch into the previous patch
+    SQUASH = 2
+
+
+@dataclass
+class Instruction:
+    patch_name: str
+    action: Action
+    apply: bool
```

Comment:
Would love to use `dataclass` and type annotations in StGit, but we are currently targeting Python 3.5 as the minimum version, so we cannot use those features yet.